src/coreclr/src/jit/hwintrinsiccodegenarm64.cpp

```diff
@@ -613,6 +613,27 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
             }
             break;
 
+            case NI_Vector64_GetElement:
+            case NI_Vector128_GetElement:
+            case NI_Vector64_ToScalar:
+            case NI_Vector128_ToScalar:
+            {
+                ssize_t indexValue = 0;
+                if (intrin.id == NI_Vector64_GetElement || intrin.id == NI_Vector128_GetElement)
+                {
+                    assert(intrin.op2->IsCnsIntOrI());
+                    indexValue = intrin.op2->AsIntCon()->gtIconVal;
+                }
+
+                // no-op if vector is float/double, targetReg == op1Reg and fetching for 0th index.
+                if (!(varTypeIsFloating(intrin.baseType) && targetReg == op1Reg && indexValue == 0))
```

Comment:
Is there a `tgtPrefUse` on ARM64 like there is on x86/x64? Should that be getting set in `lsra`?

Reply:
I don't know much about `tgtPrefUse`. @echesakovMSFT ?

Reply:
I do see `tgtPrefUse` in arm64 during `BuildHWIntrinsic` at https://github.com/dotnet/runtime/blob/master/src/coreclr/src/jit/lsraarm64.cpp#L1062 . I hope  that is enough.

Reply:
I don't think it covers this case in particular as it looks to only use `tgtPrefUse` if the instruction is `RMW`.

The simple `mov reg, reg` case wouldn't fit that and so we would need to specially handle it. 
I also don't see any handling for `Commutative` operations in which case we should be setting `tgtPrefUse2` (which I would guess also exists like it does on x86).

Reply:
I will follow-up with you offline to understand this is detail. Opened https://github.com/dotnet/runtime/issues/36454 to track the work.

Reply:
> Is there a tgtPrefUse on ARM64 like there is on x86/x64? Should that be getting set in lsra?

Yes, I enabled tgtPrefUse on Arm64 in https://github.com/dotnet/runtime/pull/34668#issue-400579467 for RMW HW intrinsics. 

IMO, if we want to add more intrinsics that set tgtPrefUse this should be done using flags in hwintrinsiclistarm64.h (e.g. HW_Flag_PreferOp1RegAsTarget) and not by manually coding each case in lsraarm64.cpp.

> I also don't see any handling for Commutative operations in which case we should be setting tgtPrefUse2 (which I would guess also exists like it does on x86).

There is not such handing yet - I don't think it is gonna be useful as on x86 - since most of the instructions are `inst tgtReg, op1Reg, op2Reg`. There is no difference between emitting this and `inst tgtReg, op2Reg, op1Reg`.

Reply:
> IMO, if we want to add more intrinsics that set tgtPrefUse this should be done using flags in hwintrinsiclistarm64.h (e.g. HW_Flag_PreferOp1RegAsTarget) and not by manually coding each case in lsraarm64.cpp.

I'd want to wait and see how many instructions actually fit this category. I'd imagine on ARM it is largely restricted to RMW (in which case the RMW flag covers it) and specially handled moves where we can make it zero cost for specific types (float/double).

On x86, we really only had to special case `FMA` (because it is a 3 op, commutative RMW instruction) and a couple of the moves.

Reply:
(Copying my comment to the issue and we can continue the conversation there 😄)